Add example tests

eric-heiden opened the pull request:
<!--
Thank you for contributing to Newton!

Please fill the relevant sections.

Checkboxes can also be marked after you submit the PR.
-->

## Description
* Implemented a `test()` function for various examples and added some helper methods to assist with verifying the state of bodies and particles
* Each example is tested for NaN values in its Model, State, Control, and Contacts structures
  * `example_sensor_contact.py` appears to have issues on CPU, I've disable the CPU version for it now and assigned https://github.com/newton-physics/newton/issues/860 to @camevor 
* Each example now executes the test function and above mentioned generic NaN check by passing `--test` to its CLI args
* Some examples were fixed, or updated to use some new API functions like `builder.replicate()`
* Some examples now only get tested on CUDA to save resources since the CPU version does not add new meaningful test coverage compared to similar examples/sim features we already test on CPU
* Note that the 2 benchmark failures are due to the slight change in API where `newton.examples.run(example, args)` now requires an additional `args` input that wasn't there before

## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this PR.

- [ ] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [ ] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [ ] Documentation is up-to-date
- [ ] Code passes formatting and linting checks with `pre-commit run -a`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added public vector helpers for elementwise closeness and in-bounds checks.
  - Examples now include built-in state tests (body/particle), test mode with NaN/non-finite checks, and loss/cost-history & gradient validation in diff‑sim examples.

- **Tests**
  - Expanded test utilities, example test registrations, per-device test options, and richer example validations.

- **Chores**
  - Runner now accepts CLI args; examples and benchmarks updated to forward args.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

adenzler-nvidia commented on newton/examples/robot/example_robot_allegro_hand.py:
```diff
@@ -196,7 +196,12 @@ def render(self):
+            self.state_0,
+            "all bodies are above the ground",
+            lambda q, qd: q[2] > 0.9,
+        )
```
is it possible to make this a bit tighter requirement, like all the body of the same hand stay within some reasonable volume? Then we can catch explosions?

eric-heiden replied:
Good idea! I've implemented this now.

adenzler-nvidia reviewed (commented):
added some comment, looking good no major objections.

2 questions:
* should we auto-discover examples and fail the tests if they don't have a test() function? We can still leave it unimplemented but it would force devs to think about whether this needs testing.
* how long are these tests taking?

eric-heiden commented on newton/_src/utils/__init__.py:
```diff
@@ -194,6 +196,28 @@ def vec_abs(a: wp.vec3):
+    for i in range(wp.static(len(a))):
+        if a[i] < lower[i] or a[i] > upper[i]:
+            return False
+    return True
```
I thought these vector-wise functions could be useful beyond testing so I added them to the utils where we already have other functions of this sort.

eric-heiden commented on newton/examples/basic/example_basic_shapes.py:
```diff
@@ -131,7 +137,46 @@ def step(self):
 
     def test(self):
-        pass
+        sphere_q = wp.transform(0.0, -2.0, 0.5, 0.0, 0.0, 0.0, 1.0)
```
Done

eric-heiden commented:
> added some comment, looking good no major objections.
> 
> 2 questions:
> 
> * should we auto-discover examples and fail the tests if they don't have a test() function? We can still leave it unimplemented but it would force devs to think about whether this needs testing.
> * how long are these tests taking?

Thanks! The auto discovery might be a bit tricky since we often need to pass custom CLI args to the examples for testing the CPU/GPU versions. I've changed it to raise a NotImplementedError now if an example doesn't have a `test()` method but is executed with a `--test` argument.

There seems to be some issue in the parallel unit test execution, it seems like we exceed 64 parallel runs now @Kenny-Vilella? I've added just a few more example tests now, maybe that caused this.

Kenny-Vilella commented:
@eric-heiden  We already saw these kind of error messages in the past.
The tests are now using 8 process in parallel, I wonder if it counts as 8 cpu, if yes then with 2 jobs per pipeline we will be limited by 4 pipelines at a time.

eric-heiden commented:
OK the tests are working now, thanks @Kenny-Vilella! 

The test time is not worse than before this PR, it seems actually slightly faster: 20min vs 22min on main.
I will go ahead and merge it then to unblock the other examples PRs that are up to leverage the test functions from here.
